Bring anti-patterns more inline with nixpkgs contributing guide (#307)
NixOS/nix.dev: source/anti-patterns/language.md

- combining with overriding logic such as `overrideAttrs` function in nixpkgs has a surprising behaviour
  of not overriding every reference.

> A better way is to use simpler but more verbose `let .. in`:

Reviewer: ```suggestion
A better way is to use `let ... in`:
```

I'm not sure in what sense `let ... in` is simpler, and I don't think it's worthwhile pointing out it requires more characters to express.

}
```

:::{note}

> It is fine to use `rec` in nixpkgs' stdenv.mkDerivation function especially for reusing version in a fetcher for src since that is a fixed output derivation anyway.

Reviewer: I can't follow the reasoning here. What makes the fixed output derivation special in this case?

> It is fine to use `rec` in nixpkgs' stdenv.mkDerivation function especially for reusing version in a fetcher for src since that is a fixed output derivation anyway.

Reviewer 2: Should be discouraged a bit more, it's not great but the best we have in some cases.

# try this instead:
let
  pkgs = import <nixpkgs> {};
  inherit (pkgs) curl jq;
in ...
```

> ```nix

Reviewer: Since I can't make suggestions on deleted lines (WTF GitHub?) and lines not modified:

l. 68:

> We recommend the following:

l. 80:

> Use `with` only for tight scopes which leave no ambiguity:
> 
> ```
> buildInputs = with pkgs; [ curl jq ];
> ```
> ```
> meta.maintainers = with lib.maintainers; [ edolstra ];
> ```

:::{note}
[Channels](https://nixos.wiki/wiki/Nix_channels) are a way of distributing Nix software.

> They are being phased out by flakes, but still used by default.

Reviewer: We should first explain visibly what flakes are and how to use them before telling people "channels are all over the place, but forget that. use flakes, but you'll have to figure out what they even do on your own."

So we would only refer to flakes once we have an authoritative (and hopefully satisfactory) definition. There are  RFC49 (closed!) and Eelco's blog posts, and I'm fine with linking those, but something with Nix™ on them would be better.

> They are being phased out by flakes, but still used by default.

Reviewer: We should first explain visibly what flakes are and how to use them before telling people "channels are all over the place, but forget that. use flakes, but you'll have to figure out what they even do on your own."

So we would only refer to flakes once we have an authoritative (and hopefully satisfactory) definition. There are  RFC49 (closed!) and Eelco's blog posts, and I'm fine with linking those, but something with Nix™ on them would be better.

I'm not even sure if the statement was true to begin with. What about a bit more cautiously saying

```suggestion
They are used by default, but there are plans to replace them with a more robust system.
```

> They are being phased out by flakes, but still used by default.

Reviewer 2: Let's not recommend flakes yet, but now we have this page giving some flakes context: https://nix.dev/concepts/flakes#flakes

2. you control the shell environment via your source code, setting `$NIX_PATH` via `nix-shell` or NixOS options.

:::{note}

> We recommend to avoid using extra search paths and to switch to flakes and disable channels by permanently setting `NIX_PATH=` to be empty.

Reviewer: Again, switching to flakes should be supported by authoritative material on how to do that, otherwise it will just confuse people.

```suggestion
We recommend to avoid using search paths and disable channels by permanently setting `NIX_PATH=` to be empty.
```

The docs team decided "[not to drag people into unstable features](https://github.com/NixOS/nix.dev/issues/281#issuecomment-1198128399)" and instead build up Nix 3 and Flakes on a separate branch.

Reviewer 2: Agreed, we shouldn't recommend flakes
